### DISCUSSION TITLE
RabbitMQ use

### DISCUSSION THREAD

**Discussion by shashikantpawar21** [Eventuous/eventuous#71] [Q&A]
Hi Alexey, 

Is the RabbitMQ eventuous stuff fully functional.
Do you have any sample code which illustrate how to use rabbitMQ in eventuous?

**Comment by alexeyzimarev** [comment]
Yes, it is. The code can be found [in the tests](https://github.com/Eventuous/eventuous/blob/011bb314305cede79e3d83277a9ffdf431844113/src/RabbitMq/test/Eventuous.Tests.RabbitMq/SubscriptionSpec.cs).

Basically, it works as a normal producer and subscription. Unfortunately, those aren't fully documented yet, but I am working on it.

**Reply by shashikantpawar21** [reply]
Thank you for quick reply. I really appreciate that.. 
I checked the test project bit difficult to understand. I will give a deep read into function and try it. 

Any sample code or documentation just like bookings project for eventuous would really help..

What i am currently trying to solve and your advice would help me ...
1. I separate edge service , aggregate domain and projections as separate microservice..
2.  I see coupling between edge service and aggregate domain as i have my command service in edge service using application service as a base. 
3. So thought of trying rabbitmq to send command from edge service to aggregate domain service.. 

Is this right approach or can u suggest something more effective to decouple dependency of edge service on aggregate domain somehow ?

**Reply by alexeyzimarev** [reply]
I don't fully follow your terminology, what's the "edge service", "aggregate domain", "aggregate domain service" are? Eventuous only has a command service (app service) and base classes for the domain model. There's an example in the Booking project of integration between bookings and payments, which uses EventStoreDB as a bus. It can easily be changed to use RMQ.

**Reply by shashikantpawar21** [reply]
With reference to Booking project example, I am trying CQRS so my query part will be separate service
So i made 3 microservice
Microservice 1 (MS1)  -->  Bookings --> httpApi (only commandApi) and command service (app service)
Microservice 2 (MS2) -->  Bookings.Domain --> (Apply data to event store db and project to mongo document)
Microservice 3 (MS3) -->  Bookings.Query --> (Read from mongo document)

 **Bookings** (MS1) depends upon **Booking.Domain**(MS2)  to know the command handler to call.. 
So, i thought of using rabbit MQ bus to publish a command from MS1 and it is will be subscribe from MS2 and call the respective command handler.. 

Is this approach right or this can be done in better way using some eventuous abstraction?

**Comment by alexeyzimarev** [comment]
The aim of the application service is to call the domain model directly. Those things belong together, and there's no sense in separating those into different executables. These things are tightly coupled.

**Reply by shashikantpawar21** [reply]
Got your point. Thank you for the help..

Just one more query...

I need to support real time processing in my application so used signalR for same. 
Now have one websocket connection to handle multiple domain request (booking and payment in reference to sample code).. 
Any suggestion how to handle this as application service is mapped to one domain ?

Any sample code with command send from signalR instead of api would be helpful

**Reply by alexeyzimarev** [reply]
Using SignalR is no different compared to HTTP API. I am not sure what the question is. You get a command in the SignalR hub and pass it over to the command service.

**Reply by shashikantpawar21** [reply]
If we see sample code ...
1. Booking related HTTP Api are inside Booking project 
2. Payment related HTTP api are inside payment project 
Both project can be scaled separately and it is nice.. 

But with Signal hub, I  have only this one project with send command to booking and payment project for processing...because cannot split signalR hub into 2,  one for booking and payment which can sits inside respective project just like HTTP Api 
Any suggestion here ?

**Reply by alexeyzimarev** [reply]
Why can't you split hubs? When you map hub endpoints, each hub can have its own path. It's totally possible to have multiple hubs that would act like command or query services.

I'd suggest that when you have a new question, you create a new discussion. It's weird that the SignalR discussion is in the RMQ topic. Thanks in advance.
